src/hyperloglog.c

```diff
@@ -664,10 +664,19 @@ int hllSparseSet(robj *o, long index, uint8_t count) {
 
     /* When updating a sparse representation, sometimes we may need to
      * enlarge the buffer for up to 3 bytes in the worst case (XZERO split
-     * into XZERO-VAL-XZERO). Make sure there is enough space right now
-     * so that the pointers we take during the execution of the function
-     * will be valid all the time. */
-    o->ptr = sdsMakeRoomFor(o->ptr,3);
+     * into XZERO-VAL-XZERO).
+     * And we should not allocate more than server.hll_sparse_max_bytes for
+     * the sparse representation. So we resize it if the memory reaches its
+     * max size.
+     * If the available size of hyperloglog sds string is not enough for the
+     * increment we need, we promote the hypreloglog to dense representation. 
+     */
+    if (sdsalloc(o->ptr) < server.hll_sparse_max_bytes) {
+        o->ptr = sdsMakeRoomFor(o->ptr,3);
+        if (sdsalloc(o->ptr) > server.hll_sparse_max_bytes) {
+            o->ptr = sdsResize(o->ptr, server.hll_sparse_max_bytes);
+        }
```

Comment:
we now have `sdsMakeRoomForNonGreedy`, let's use it, or we can just use `sdsRezise`.
the current approach of calling sdsMakeRoomFor and then sdsRezise is wasteful since it can copy the data twice.

Reply:
I think it still meaningful to use the greedy function, as we don't want to enlarge the memory again and again frequently.
We can predict whether it will exceed the `server.hll_sparse_max_bytes` before we chose to use `sdsMakeRoomFor` or `sdsResize`

Reply:
ok, but i don't like the fact the code assumes what sdsMakeRoomFor does.
maybe just fully implement the greediness in hll using sdsResize

Reply:
ok, updated.